tcpproxy.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package testing
+
+import (
+	"io"
+	"net"
+	"sync"
+
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+)
+
+// TCPProxy is a simple TCP proxy that can be used
+// to deliberately break TCP connections.
+type TCPProxy struct {
+	listener net.Listener
+	// mu guards the fields below it.
+	mu sync.Mutex
+	// closed holds whether the proxy has been closed.
+	closed bool
+	// conns holds all connections that have been made.
+	conns []io.Closer
+}
+
+// NewTCPProxy runs a proxy that copies to and from
+// the given remote TCP address. When the proxy
+// is closed, its listener and all connections will be closed.
+func NewTCPProxy(c *gc.C, remoteAddr string) *TCPProxy {
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	c.Assert(err, jc.ErrorIsNil)
+	p := &TCPProxy{
+		listener: listener,
+	}
+	go func() {
+		for {
+			client, err := p.listener.Accept()
+			if err != nil {
+				if !p.isClosed() {
+					c.Error("cannot accept: %v", err)
+				}
+				return
+			}
+			p.addConn(client)
+			server, err := net.Dial("tcp", remoteAddr)
+			if err != nil {
+				if !p.isClosed() {
+					c.Error("cannot dial remote address: %v", err)
+				}
+				return
+			}
+			p.addConn(server)
+			go stream(client, server)
+			go stream(server, client)
+		}
+	}()
+	return p
+}
+
+func (p *TCPProxy) addConn(c net.Conn) {
+	p.mu.Lock()
```

Comment:
+1 on this too. The argument that the function is trivial won't work on me either, because functions change